### PR TITLE
[DataGrid] Preprocess edit cell props on backspace/delete

### LINKED ISSUE #15058: [data grid] Trigger `preProcessEditCellProps` from view by e.g onKeyDown / similar to e.g paste from clipboard (for date fields)
### Description

We use `preProcessEditCellProps` for validating entries in cells (mostly in `editMode === "row"`).

In some use cases we need to validate date fields (e.g if the value is after provided date or `null` is not allowed). It works also almost out of the box, by using `GridDataEditCell` with a custom `Tooltip`, but only, if the values was set over built-in Date picker.

Is there a way to trigger / start validation (aka `preProcessEditCellProps`) by e.g entering the edit mode? Some use cases:
- selecting the cell and pressing `Backspace` (or starting to type date eg. "202310" to remove the value should trigger validation
- selecting the cell and pasting from clipboard the value (e.g "2023-10-30"), should also trigger validation process

### Snippet examples

```tsx
const preProcessEditCellProps = (props: GridPreProcessEditCellProps) => {
  let error;
  let errorText = '';

  const { value } = params.props;

  if (compareDate) {
    if (!value) {
      error = true;
      errorText = "this field is required";
    }
    if (value && !isAfter(new Date(value), compareDate)) {
      error = true;
      errorText = 'The date should be after ${compareDate}'; // left formatiing code here
    }
  }

  return { ...params.props, error, errorText };
}

...
const columns = [
 {
      field: 'compareDate',
      type: 'date',
      editable: true,
      headerName: 'Compare date,
      disableColumnMenu: true,
      flex: 2,
      renderEditCell: renderDateCell, // basically the GridEditDateCell with a tooltip
      valueFormatter: gridBasicDateFormatter,
      preProcessEditCellProps: (params: GridPreProcessEditCellProps) => {
        return preProcessVersionVonAvailable(params, compareDate);
      }
    },
...
}

....

<DataGridPro
  columns={columns}
  ...
/>

```

P.S: I've also noticed that validation is being triggered, **ONLY** when the full date was entered (via keyboard or selected from the date picker). Is it possible to adjust this behaviour and run it e.g for not fully entered date (via keyboard)?

Look forward to hearing from you!

### Your environment

<details>
  <summary>`npx @mui/envinfo`</summary>

```
  Don't forget to mention which browser you used.
  Output from `npx @mui/envinfo` goes here.
```
</details>

**Search keywords**: preProcessEditCellProps validation Date clipboard
**Order ID**: 62555

### PR DESCRIPTION
**Pull request by KenanYusuf**
Calls `preProcessEditCellProps` when a cell contents is deleted by the backspace/delete key.

Before: https://codesandbox.io/p/sandbox/15151-preprocess-edit-cell-on-delete-after-forked-7vwlyg
After: https://codesandbox.io/p/sandbox/elegant-pond-nchrgq

Closes #15058

### CODE REVIEW

**Comment by arminmeh** [comment]
Can you add a test for this?

**Comment by KenanYusuf** [comment]
> Can you add a test for this?

@arminmeh I have added a basic test for cell and row editing - anything else you think should be tested?

**Comment by arminmeh** [comment]
> I have added a basic test for cell and row editing - anything else you think should be tested?

Ignore 😆
Collapsed the files when I started. Continued review today, forgot about the test being there 😅

**Review by arminmeh — APPROVED** [review]

File: packages/x-data-grid/src/hooks/features/editing/useGridCellEditing.ts [hunk 1]
```diff
@@ -333,25 +333,49 @@ export const useGridCellEditing = (
...
+      let newValue = value;
       if (deleteValue) {
         newValue = getDefaultCellValue(apiRef.current.getColumn(field));
       } else if (initialValue) {
         newValue = initialValue;
       }
 
-      const newProps = {
+      const column = apiRef.current.getColumn(field);
+      const shouldProcessEditCellProps = !!column.preProcessEditCellProps && deleteValue;
```
[the remark this answers is not in the document]
Comment (KenanYusuf) [reply packages/x-data-grid/src/hooks/features/editing/useGridCellEditing.ts:348]: It's a good point - I did have it like that initially, but it felt like a slightly strange UX to see the cell processing as soon as I entered into it. I think it should probably be down to the developer to ensure that the values that populate the grid comply with their own validation rules. WDYT?

[the remark this answers is not in the document]
Comment (arminmeh) [reply packages/x-data-grid/src/hooks/features/editing/useGridCellEditing.ts:348]: I guess we can keep it like this for now and update if there is a request for it

**Pull request merged**